src/main/java/com/metamx/common/FileUtils.java

```diff
@@ -0,0 +1,291 @@
+   *
+   * @param inputStreamSupplier  Supplier for input streams to copy from. The stream is closed on every retry.
+   * @param shouldRetry          Predicate to determine if the throwable is recoverable for a retry
+   * @param outputStreamSupplier Supplier for output streams. The stream is closed on every retry.
+   * @param maxAttempts          Maximum number of retries before failing
+   */
+  public static long retryCopy(
+      final ByteSource inputStreamSupplier,
+      final Predicate<Throwable> shouldRetry,
+      final ByteSink outputStreamSupplier,
+      final int maxAttempts
+  )
+  {
+    try {
+      return RetryUtils.retry(
+          new Callable<Long>()
+          {
+            @Override
+            public Long call() throws Exception
+            {
+              InputStream inputStream = null;
+              OutputStream outputStream = null;
+              try {
+                inputStream = inputStreamSupplier.openStream();
+                outputStream = outputStreamSupplier.openStream();
+                return ByteStreams.copy(inputStream, outputStream);
+              }
+              finally {
+                CloseQuietly.close(inputStream);
+                CloseQuietly.close(outputStream);
+              }
+            }
+          },
+          shouldRetry,
+          maxAttempts
+      );
+    }
+    catch (Exception e) {
+      throw Throwables.propagate(e);
+    }
+  }
+
+  public static class MD5MismatchException extends IOException
+  {
+    public MD5MismatchException(String msg)
+    {
+      super(msg);
+    }
+
+    public MD5MismatchException(String msg, String... args)
+    {
+      super(String.format(msg, args));
+    }
+  }
+
+  public static class FileCopyResult
+  {
+    private final Collection<FileCopyResult> children = Lists.newArrayList();
+    private final Path path;
+    private volatile long childSize = -1l;
+    private volatile long lastSizeCheck = 0l;
+    private volatile long childModCount = 0l;
+    private final MessageDigest digest;
+    private final byte[] digestBytes;
+
+    public Collection<FileCopyResult> getChildren()
+    {
+      return ImmutableList.copyOf(children);
+    }
+
+    public MessageDigest getDigest()
+    {
+      return digest;
+    }
+
+    public Path getPath()
+    {
+      return path;
+    }
+
+    // Consumes otherDigest.digest()
+    public Boolean digestEquals(MessageDigest otherDigest){
+      if(this.digest == null || this.digestBytes == null){
+        return false;
+      }
+      if(!otherDigest.getAlgorithm().equals(digest.getAlgorithm())){
+        throw new IAE("Digest algorithm did not match. Expected [%s] got [%s]", digest.getAlgorithm(), otherDigest.getAlgorithm());
+      }
+      return Arrays.equals(digestBytes, otherDigest.digest());
+    }
+
+    // Only works for immutable children contents
+    public synchronized Long size()
```

Comment:
Doing this simply will cause two major changes:
1. No longer thread safe
2. Tree must be built from the leaves up

I think these are acceptable limitations given the use case. I can add that to the description.